doc/designs/disable-stale-users.md

```diff
@@ -0,0 +1,273 @@
+# Disable Stale Users
+
+**DESIGN STAGE**
+
+## Overview
+
+Some regulations call for accounts that have not been used for a specific
+amount of time be automatically disabled in the identity management system.
+
+FreeIPA does not keep track of successful authentication events.
+The krbLastSuccessfulAuth attribute is not updated by default since
+https://pagure.io/freeipa/issue/5313 for reliability reasons.
+Therefore distinguishing between stale and active accounts cannot be done with
+existing logon data.
+
+
+## Possible approaches
+
+### Environments with mandatory, periodic password changes
+
+The problem is easy to solve in environments where user passwords must be
+changed regularly. A user account whose associated password is past its due
+change date can be considered stale and therefore be disabled by the system.
+In this case, a grace period must be provided between password expiration and
+the actual account locking event.
+Total writes to LDAP using this approach due to the disable stale users tooling
+are minimal (only nsAccountLock). However, password-change related writes are
+needed every 90 days (by default).
+Each password change updates the krbLastPwdChange, krbPasswordExpiration,
+userPassword, krbPrincipalKey, and krbExtraData attributes.
+
+The Disable Stale Users (DSU) tooling proposes to make sure no non-locked
+account has a password expired since N days, N being 7 by default.
+
+### Environments with no mandatory password changes
+
+For environments where user passwords never expire, a new mechanism must be
+added to FreeIPA plugins ipa-kdb and ipa-lockout to account for each user
+account authentication events. Since updating a LDAP user account attribute on
+each authentication event must be avoided as the update frequency could easily
+be high, (see https://pagure.io/freeipa/issue/5313 for background
+information), a coarse update mechanism is proposed.
+
+An attribute containing a timestamp is added to each user:
+coarseSuccessfulAuthTime. The timestamp contained therein is not precise.
+It will be added at account creation time (set to current date).
+The timestamp is updated by IPA plugins whenever a successful
+authentication event happens, but only if the current timestamp is
+sufficiently in the past. The actual 
+
+Some attributes must be added to the cn=ipaConfig schema:
+* the first one to activate the mechanism (default: not present):
+  ipaConfigString: DSU:RecordSuccessfulAuth
+* the second attribute contains the fixed time period (default: 28 days):
+  ipaDSURecordSucessfulAuthTimeInterval: 2419200
+* the third one contains a random time period to be added. The actual value
+  used is randomized between 0 and the attribute's value (default: 7 days)
+  whenever a user successfully authenticates. 
+  ipaDSURecordSucessfulAuthRandomizedTimeInterval: 604800
+
+The time period to take into account is 28+7 so 35 days.
```

Comment:
It's range of 28-35 days, right?

It's a bit unclear how to formulate this as both 61 and 55 days are mentioned. 55 caps at 90 days so i assume that's what you mean. But this means some accounts could be disabled early.

28 + 55 = 83
29 + 55 = 84
30 + 55 = 85
...
35 + 55 = 90

So the range is 83-90 based on the randomized time interval. The user could configure that shorter or longer at the cost of potentially more accounts needing updates at the same time. The bigger the spread the more spread out writes are.

28 + 61 = 89
29 + 61 = 90
...
35 + 61 = 96

The text is reasonable but for quick reads it would help IMHO to include the formula for determining the expiration.

The days constant is 90 - coarseSuccessfulAuthTime + ipaDSURecordSucessfulAuthRandomizedTimeInterval

You also need an attribute to store the days-until-expiration value (90 by default).

Reply:
re: "You also need an attribute to store the days-until-expiration value (90 by default)."
=> This is explained later, see TimeStampBasedDSUInactivePeriod.

Reply:
Q: "It's range of 28-35 days, right?"
A: yes. I've added "Maximum" to the above sentence

Reply:
I'm adding a formula.